src/Bicep.Core/Modules/TemplateSpecModuleReference.cs

```diff
@@ -0,0 +1,114 @@
+
+        private static readonly Regex ResourceNameRegex = new(@"^[-\w\._\(\)]{0,89}[-\w_\(\)]$");
+
+        private TemplateSpecModuleReference(string? endpoint, string subscriptionId, string resourceGroupName, string templateSpecName, string version)
+            : base(ModuleReferenceSchemes.TemplateSpecs)
+        {
+            this.Endpoint = endpoint;
+            this.SubscriptionId = subscriptionId;
+            this.ResourceGroupName = resourceGroupName;
+            this.TemplateSpecName = templateSpecName;
+            this.Version = version;
+        }
+
+        public override string UnqualifiedReference => this.Endpoint is not null
+            ? $"{this.Endpoint}/{this.SubscriptionId}/{this.ResourceGroupName}/{this.TemplateSpecName}:{this.Version}"
+            : $"{this.SubscriptionId}/{this.ResourceGroupName}/{this.TemplateSpecName}:{this.Version}";
+
+        public string? Endpoint { get; }
+
+        public string SubscriptionId { get; }
+
+        public string ResourceGroupName { get; }
+
+        public string TemplateSpecName { get; }
+
+        public string Version { get; }
+
+        public Uri? EndpointUri => this.Endpoint is not null ? new Uri($"https://{this.Endpoint}") : null;
+
+        public string TemplateSpecResourceId => $"/subscriptions/{this.SubscriptionId}/resourceGroups/{this.ResourceGroupName}/providers/Microsoft.Resources/templateSpecs/{this.TemplateSpecName}/versions/{this.Version}";
+
+        public override bool Equals(object? obj) =>
+            obj is TemplateSpecModuleReference other &&
+            StringComparer.OrdinalIgnoreCase.Equals(this.Endpoint, other.Endpoint) &&
+            StringComparer.OrdinalIgnoreCase.Equals(this.SubscriptionId, other.SubscriptionId) &&
+            StringComparer.OrdinalIgnoreCase.Equals(this.ResourceGroupName, other.ResourceGroupName) &&
+            StringComparer.OrdinalIgnoreCase.Equals(this.TemplateSpecName, other.TemplateSpecName) &&
+            StringComparer.OrdinalIgnoreCase.Equals(this.Version, other.Version);
+
+        public override int GetHashCode()
+        {
+            var hash = new HashCode();
+
+            hash.Add(this.Endpoint, StringComparer.OrdinalIgnoreCase);
+            hash.Add(this.SubscriptionId, StringComparer.OrdinalIgnoreCase);
+            hash.Add(this.ResourceGroupName, StringComparer.OrdinalIgnoreCase);
+            hash.Add(this.TemplateSpecName, StringComparer.OrdinalIgnoreCase);
+            hash.Add(this.Version, StringComparer.OrdinalIgnoreCase);
+
+            return hash.ToHashCode();
+        }
+
+        public static TemplateSpecModuleReference? TryParse(string value, out DiagnosticBuilder.ErrorBuilderDelegate? failureBuilder)
+        {
+            static DiagnosticBuilder.ErrorBuilderDelegate CreateErrorFunc(string rawValue) => x => x.InvalidTemplateSpecReference($"oci:{rawValue}");
```

Comment:
>oci [](http://example.com/codeflow?start=132&length=3)

:)

Reply:
Should also use the schemes constants. The value is just ability to quickly find things that depend on it though. We could never change it after we ship.